[BFW-7076] gcode: rework M865

bkerler opened the pull request:
This solves issue #4542 by adding an option to print the filament type for a specific tool slot:

`M865 I0` will print for tool slot 0 (Extruder 1):

![image](https://github.com/user-attachments/assets/997f5170-7361-4c21-b621-8b1d20e7b8c6)

Furthermore:
- Rename F<type> to S<type> due to conflict with F (Required filtration), update documentation
- Fix max value of Extruder slot for L<ix>, update documentation
- Change behavior of I<ix> for changing/getting the current filament, update documentation

CZDanol commented on src/marlin_stubs/M865.cpp:
```diff
@@ -36,6 +37,10 @@ void PrusaGcodeSuite::M865() {
     }
 
     FilamentType filament_type;
+    if (const auto slot = p.option<uint8_t>('H', static_cast<uint8_t>(0), static_cast<uint8_t>(adhoc_filament_type_count - 1))) {
```
`adhoc_filament_type_count` is not the right variable to check against here. You wanna do `EXTRUDERS`

CZDanol commented on src/marlin_stubs/M865.cpp:
```diff
@@ -25,6 +25,7 @@
  * - `A` - Is abrasive
  * - `F` - Requries filtration
  * - `N"<string>"` - New filament name
+ * - `H`<i>` - Print current filament type loaded to the specified tool (indexed from 0)
```
˙H˙ is already taken in our trunk code unfortunately

CZDanol reviewed (changes requested):
While the idea is noble, the implementation is rather unconceptual. Let's try to redesign this feature in the spirit of this gcode:

1. Add a "selector" for a filament loaded in a specific tool. There is already the `I<ix>` for the "Custom" filament type in a specified tool - we could simply change that to the current filament loaded to the specified tool, only adding functionality
2. Do we need a print option? Why not print things about the filaments always, in a parseable manner? And why limit ourselves only to the filament name?

CZDanol commented on src/marlin_stubs/M865.cpp:
```diff
@@ -51,7 +48,6 @@ void PrusaGcodeSuite::M865() {
 
     } else if (const auto ft = p.option<FilamentType>('F')) {
         filament_type = *ft;
-
```
Please keep this line

bkerler replied:
Commit format tool doesn't like this line :)

CZDanol commented on src/marlin_stubs/M865.cpp:
```diff
@@ -37,12 +38,8 @@ void PrusaGcodeSuite::M865() {
-            SERIAL_ERROR_MSG("The selected tool does not have the ad-hoc filament loaded. Changes will have no effect.");
-        }
-
+    if (const auto slot = p.option<uint8_t>('I', static_cast<uint8_t>(0), static_cast<uint8_t>(EXTRUDERS))) {
```
We also need to change the documentation

CZDanol commented on src/marlin_stubs/M865.cpp:
```diff
@@ -37,12 +38,8 @@ void PrusaGcodeSuite::M865() {
-            SERIAL_ERROR_MSG("The selected tool does not have the ad-hoc filament loaded. Changes will have no effect.");
-        }
-
+    if (const auto slot = p.option<uint8_t>('I', static_cast<uint8_t>(0), static_cast<uint8_t>(EXTRUDERS - 1))) {
```
The documentation for this parameter should also be updated

CZDanol replied:
Changing behavior (and documentation) of the I parameter should be a separate commit

CZDanol commented on src/marlin_stubs/M865.cpp:
```diff
@@ -1,4 +1,5 @@
 #include "PrusaGcodeSuite.hpp"
+#include "marlin_vars.hpp"
```
We don't need this include

CZDanol commented on src/marlin_stubs/M865.cpp:
```diff
@@ -11,6 +11,7 @@
  *
  *#### Parameters
  * - `I<ix>` - Configure parameters of a Custom filament currently loaded to the specified tool (indexed from 0)
+ * - `I<i>`  - Show the parameters of the filament currently loaded to the specified tool (indexed from 0)
```
- The rest is using the "Configure parameters" text. I'd suggest either updating all texts, or keeping it unified.
- The rest is using "ix", not "i"
- Also, you've duplicated the documentation for I.

bkerler committed:
Rework I option in M865 gcode

bkerler committed:
Add printing of parameters in M865 gcode

bkerler committed:
Change already used F option to S in M865 gcode.
Used to be F"<preset>" in the past and was renamed to S"<preset>" due to conflict with option "F" for filtration required.

bkerler committed:
M865 fix EXTRUDER param max size for L option

CZDanol commented on src/marlin_stubs/M865.cpp:
```diff
@@ -28,6 +28,7 @@
  *
  * Ad-hoc/custom filaments can the be referenced in other gcodes using adhoc_filament_gcode_prefix.
  * For example `M600 S"#0"` will load ad-hoc filament previously set with `M865 I0`.
+ * The current filament settings of the specified tool (indexed from 0) are printed to the serial console
```
>  The current filament settings of the specified tool
Only if you use the `I` selector. Is not true if you use any other selector.

CZDanol commented on src/marlin_stubs/M865.cpp:
```diff
@@ -13,7 +13,7 @@
  * - `I<ix>` - Configure parameters of a filament currently loaded to the specified tool (indexed from 0)
  * - `U<ix>` - Configure parameters of a User filament (indexed from 0)
  * - `X` - Configure parameters of a Custom filament type that will be loaded using `M600 F"##"` (or similar filament change gcode)
- * - `F"<preset>"` - Configure parameters of User filament with this name  (or select Preset filament for `L`)
```
This commit needs a description explaining why this change was needed.

CZDanol commented on src/marlin_stubs/M865.cpp:
```diff
@@ -13,8 +13,7 @@
  * - `X` - Configure parameters of a Custom filament type that will be loaded using `M600 F"##"` (or similar filament change gcode)
- * - `F"<preset>"` - Configure parameters of User filament with this name  (or select Preset filament for `L`)
- *
+ * - `S"<preset>"` - Configure parameters of filament with this name  (or select Preset filament for `L`)
```
Nitpick: You left `<preset>` here - should be `<name>` now

bkerler replied:
Done

CZDanol commented on src/marlin_stubs/M865.cpp:
```diff
@@ -23,12 +22,13 @@
  * Ad-hoc/custom filaments can the be referenced in other gcodes using adhoc_filament_gcode_prefix.
  * For example `M600 S"#0"` will load ad-hoc filament previously set with `M865 I0`.
- * The current filament settings of the specified tool (indexed from 0) are printed to the serial console
+ * The filament_type settings are printed to the serial console. For option I, the current filament_type
```
Nitpick: Why do you keep referring to the `I` parameter here, like it's something special?

bkerler replied:
Correct, removed that.

CZDanol commented:
I have now checked and the buffer size you selected in `ArrayStringBuilder<filament_name_buffer_size + 82> filament_info;` doesn't even cover the key names length. Rather than going another review round, I will fix this privately and close this PR.

CZDanol reviewed (changes requested):
I have now checked and the buffer size you selected in ArrayStringBuilder<filament_name_buffer_size + 82> filament_info; doesn't even cover the key names length. Rather than going another review round, I will fix this privately and close this PR.

bkerler commented:
Meanwhile I've updated the PR following your comments and tested it as well :)

CZDanol commented:
Well, I have fixed it and merged in our codebase a month ago, so that was pointless. I just forgot to close the PR here - so doing it now. Sorry.
